[testing] [ci] Execute module and matrix tests before build to unify local and GHA testing

gmaiorov opened the pull request:
## Description
<!---
  Describe your changes in detail.

  Please let users know if your feature influences critical cluster components
  (restarts of ingress-controllers, control-plane, Prometheus, etc).
-->

This PR does following:
1. Patches module tests for the following modules to allow testing before Werf merges CE/EE/FE directories:
- `admission-policy-engine`;
- `node-manager`;
- `user-authz`;
- `cloud-provider-openstack`;
- `cloud-provider-vsphere`;
- `istio`;
2. Changes hooks in `monitoring-application` and `istio` modules to correctly address `D8_IS_TESTS_ENVIRONMENT` environment variable;
2. Removes module tests for `keepalived` module;
3. Changes `build-and-test_dev.yml`, `build-and-test_pre-release.yml`, `build-and-test_release.yml` workflows to perform module and matrix testing independently of build via same `Makefile` target that is used for local testing;
4. Fixes following bugs in focused module testing via `Makefile`:
- `make FOCUS=prometheus tests-modules` matches both `prometheus` and `operator-prometheus` modules;
- `make FOCUS=<module> tests-modules` breaks if module is located in more than one of the `ce`/`ee`/`fe` directories.

Testing was done in `test-1` repository. Runs corresponding to changed workflows available at:
* [`build-and-test_dev.yml`](https://github.com/deckhouse/deckhouse-test-1/actions/runs/4324218390);
* [`build-and-test_pre-release.yml`](https://github.com/deckhouse/deckhouse-test-1/actions/runs/4324227466);
* [`build-and-test_release.yml`](https://github.com/deckhouse/deckhouse-test-1/actions/runs/4324202680).

## Why do we need it, and what problem does it solve?
<!---
  This is the most important paragraph.
  You must describe the main goal of your feature.

  If it fixes an issue, place a link to the issue here.

  If it fixes an obvious bug, please tell users about the impact and effect of the problem.
-->

Current testing process is designed for directory structure created after Werf editions merge.
This leads to issues with non-uniformity between local testing and testing via Github Actions, and inability to run or debug some module tests locally.

## What is the expected result?
<!---
  How can one check these changes after applying?  

  Describe, what (resource, state, event, etc.) MUST or MUST NOT change/happen after applying these changes.
-->
PR solves the issues above.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries
<!---
  Describe the changes so they will be included in a release changelog.

  Find examples and documentation below, or visit the instruction page on the repo wiki
  https://github.com/deckhouse/deckhouse/wiki/How-to-add-to-changelog
-->

```changes
section: testing
type: feature
summary: [testing] [ci] Execute module and matrix tests before build to unify local and GHA testing
impact_level: default
```

<!---
`impact_level: default` adds to changelog as usual, this is the default that can be omitted
`impact_level: high`    something important for users, the impact will be copied to "Know Before Update" section
`impact_level: low`     omitted in changelog YAML; note there is `type:chore` for chores

Tip for the section field:

  - <kebab-case of a module>, e.g. "cloud-provider-aws", "node-manager"
  - "ci", has forced low impact
  - "docs", includes website changes, should have low impact
  - "candi"
  - "deckhouse-controller"
  - "dhctl"
  - "global-hooks"
  - "go_lib"
  - "helm_lib"
  - "jq_lib"
  - "shell_lib"
  - "testing", has forced low impact
  - "tools", has forced low impact

Find changed sections:

gh pr diff   $PULL_REQUEST_NUMBER   |
  egrep "^([+]{3} b|[-]{3} a)/" |
  cut -d/ -f2- |
  sed 's#^ee/##' |
  sed 's#^fe/##' |
  sed 's#^modules/##' |
  sed 's#[0-9][0-9][0-9]-##' |
  egrep -v 'Makefile' |       # add file exclusion here
  cut -d/ -f1 |
  sort |
  uniq

Find all possible sections (excluding ci):

node -e 'console.log(require("./.github/scripts/js/changelog-find-sections.js")().join("\n"))'
-->

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/static'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/vsphere'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/yandex-cloud'...
<!-- workflow_start -->

gmaiorov commented:
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 4074459948 4074459948-1-con-1-25 /sys/deckhouse-oss/install:pr3740 cloud-user@51.250.67.197

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've recognized your '/e2e/abort' command and started the workflow...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

RomanenkoDenys commented on Makefile:
```diff
@@ -50,9 +50,21 @@ endif
+	TESTS_PATH = ./modules/... ./global-hooks/... ./ee/modules/... ./ee/fe/modules/...
 else
-       TESTS_PATH = $(wildcard ./modules/*-${FOCUS} ./ee/modules/*-${FOCUS} ./ee/fe/modules/*-${FOCUS})/...
+	CE_MODULES = $(shell find ./modules/ -maxdepth 1 -regextype sed -regex ".*[0-9]\{3\}-${FOCUS}")
```
shell find ./modules/ -maxdepth 1 -regextype sed -regex ".*[0-9]\{3\}-${FOCUS}" cannot work on mac

apolovov replied:
:(
```
find: -regextype: unknown primary or operator
```

gmaiorov replied:
I've addressed this issue in https://github.com/deckhouse/deckhouse/pull/3740/commits/ecf000f89a99d840a6578e65e2d17c59a0bcb194.

RomanenkoDenys commented on ee/modules/110-istio/hooks/ensure_crds_istio.go:
```diff
@@ -52,5 +53,9 @@ func ensureCRDs(input *go_hook.HookInput, dc dependency.Container) error {
 	sort.Sort(semver.Collection(semvers))
 
 	CRDversionToInstall := fmt.Sprintf("%d.%d", semvers[len(semvers)-1].Major(), semvers[len(semvers)-1].Minor())
+
+	if os.Getenv("D8_IS_TESTS_ENVIRONMENT") != "" {
+		return ensure_crds.EnsureCRDsHandler("/deckhouse/ee/modules/110-istio/crds/istio/"+CRDversionToInstall+"/*.yaml")(input, dc)
+	}
```
use variable please like an another modules

gmaiorov replied:
I've refactored this condition in https://github.com/deckhouse/deckhouse/pull/3740/commits/76e1162876c282fbd869bd1fae39cbbdfcadbc17.

RomanenkoDenys commented on modules/040-node-manager/hooks/machineclass_checksum_assign.go:
```diff
@@ -165,10 +165,18 @@ func readChecksumTemplate(cloudType string) ([]byte, error) {
 }
 
 func getChecksumTemplatePath(cloudType string) string {
+	const modulesDirCE = "/deckhouse/modules"
+	const modulesDirEE = "/deckhouse/ee/modules"
+	const modulesDirFE = "/deckhouse/ee/fe/modules"
```
Why we need this const?

gmaiorov replied:
I've inlined this constants in https://github.com/deckhouse/deckhouse/pull/3740/commits/5fef301bb025fabb248f6d4fab7fc49136fd8485.

RomanenkoDenys commented on modules/040-node-manager/hooks/machineclass_checksum_assign_test.go:
```diff
@@ -941,12 +941,26 @@ func newCloudProviderAvailabilityChecker() func(tYpE string) {
 // containing corresponding checksum template in cloud-providers directory.
 func getAvailableCloudProviderTypes() set.Set {
-	ptypes := set.New()
+	const modulesDirCE = "/deckhouse/modules"
+	const modulesDirEE = "/deckhouse/ee/modules"
+	const modulesDirFE = "/deckhouse/ee/fe/modules"
```
same question

gmaiorov replied:
I've inlined this constants in https://github.com/deckhouse/deckhouse/pull/3740/commits/5fef301bb025fabb248f6d4fab7fc49136fd8485.

RomanenkoDenys commented on modules/040-node-manager/template_tests/module_test.go:
```diff
@@ -507,6 +508,20 @@ internal:
 	f := SetupHelmConfig(``)
 
+	BeforeSuite(func() {
+		err := os.Symlink("/deckhouse/ee/modules/030-cloud-provider-openstack/cloud-instance-manager", "/deckhouse/modules/040-node-manager/cloud-providers/openstack")
```
can we define global constants with paths ?

gmaiorov replied:
I've extracted CIMs' paths to a global constants in https://github.com/deckhouse/deckhouse/pull/3740/commits/155087fb679297a6a3a51171691e704fd03f7086.

RomanenkoDenys reviewed (commented):
some nits

apolovov commented on ee/modules/110-istio/hooks/ensure_crds_istio.go:
```diff
@@ -52,5 +53,10 @@ func ensureCRDs(input *go_hook.HookInput, dc dependency.Container) error {
 	CRDversionToInstall := fmt.Sprintf("%d.%d", semvers[len(semvers)-1].Major(), semvers[len(semvers)-1].Minor())
-	return ensure_crds.EnsureCRDsHandler("/deckhouse/modules/110-istio/crds/istio/"+CRDversionToInstall+"/*.yaml")(input, dc)
+
+	editionPrefix := "/deckhouse/"
```
just prefix pls.

gmaiorov replied:
Renamed in https://github.com/deckhouse/deckhouse/pull/3740/commits/37d608b4cb0d9143d5c3eede47e2f2af552a2433.

apolovov commented:
The next step — to get rid of tests image building.

diafour commented on .github/ci_templates/tests.yml:
```diff
@@ -1,7 +1,7 @@
-args: 'go test -timeout=${{env.TEST_TIMEOUT}} -vet=off ./modules/... ./global-hooks/...'
-docker_options: '-w /deckhouse -v ~/go-pkg-cache:/go/pkg'
+args: 'make tests-modules'
+docker_options: '-w /deckhouse -v ${{github.workspace}}:/deckhouse -e "TERM=xterm-256color" -e "CGO_ENABLED=0" -v ~/go-pkg-cache:/go/pkg'
```
Why CGO_ENABLED=0? It should build with default CGO_ENABLED=1 without gcc and jg.h since shell-operator 1.1.0

diafour commented on werf.yaml:
```diff
@@ -726,6 +726,9 @@ import:
   after: setup
+shell:
+  setup:
+    - ln -s /deckhouse/modules/images_tags.json /deckhouse/modules/040-node-manager/images_tags.json
```
It seems we can just import images_tags.json one more time in the `import` section.

gmaiorov commented:
- Rebased onto `main` due to merge conflicts;
- Addressed issues brought up by @diafour;
- Updated test runs on `test-1` repo in PR description correspondingly.

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

gmaiorov commented:
Rebased due to merge conflict.

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->

gmaiorov commented:
/e2e/abort gcp;WithoutNAT;containerd;1.23 4296787637 4296787637-1-con-1-23 /sys/deckhouse-oss/install:pr3740 user@34.141.21.34

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've recognized your '/e2e/abort' command and started the workflow...
<!-- workflow_start -->

gmaiorov commented:
Rebased due to e2e fix in main

deckhouse-BOaTswain commented:
Aye, aye, @gmaiorov. I've started the workflow for label 'e2e/run/gcp'...
<!-- workflow_start -->
